src/CoreWCF.NetTcp/src/CoreWCF/Channels/Framing/ExceptionConvertingDuplexPipe.cs

```diff
@@ -0,0 +1,195 @@
+                        ConvertReceiveException(socketException));
+                }
+            }
+
+            public override bool TryRead(out ReadResult result) => _input.TryRead(out result);
+
+            Exception ConvertReceiveException(SocketException socketException)
+            {
+                return ConvertTransferException(socketException, socketException, aborted: false);
+            }
+        }
+
+        private class ExceptionConvertingPipeWriter : PipeWriter
+        {
+            private PipeWriter _output;
+
+            public ExceptionConvertingPipeWriter(PipeWriter output) => _output = output;
+
+            public override void Advance(int bytes) => _output.Advance(bytes);
+            public override void CancelPendingFlush() => _output.CancelPendingFlush();
+            public override void Complete(Exception exception = null) => _output.Complete(exception);
+            public override async ValueTask<FlushResult> FlushAsync(CancellationToken cancellationToken = default)
+            {
+                try
+                {
+                    return await _output.FlushAsync(cancellationToken);
+                }
+                catch (SocketException socketException)
+                {
+                    throw DiagnosticUtility.ExceptionUtility.ThrowHelper(
+                        ConvertSendException(socketException));
+                }
+            }
+
+            public override Memory<byte> GetMemory(int sizeHint = 0) => _output.GetMemory(sizeHint);
+            public override Span<byte> GetSpan(int sizeHint = 0) => _output.GetSpan(sizeHint);
+
+            Exception ConvertSendException(SocketException socketException)
+            {
+                return ConvertTransferException(socketException, socketException, aborted: false);
+            }
+        }
+
+        static Exception ConvertTransferException(SocketException socketException, Exception originalException, bool aborted)
+        {
+            if (socketException.ErrorCode == UnsafeNativeMethods.ERROR_INVALID_HANDLE)
+            {
+                return new CommunicationObjectAbortedException(socketException.Message, socketException);
+            }
+
+            if (socketException.ErrorCode == UnsafeNativeMethods.WSAENETRESET ||
+                socketException.ErrorCode == UnsafeNativeMethods.WSAECONNABORTED ||
+                socketException.ErrorCode == UnsafeNativeMethods.WSAECONNRESET)
+            {
+                if (aborted)
+                {
+                    return new CommunicationObjectAbortedException(SR.TcpLocalConnectionAborted, originalException);
+                }
+                else
+                {
+                    CommunicationException communicationException = new CommunicationException(SR.Format(SR.TcpConnectionResetError, "unknown"), originalException);
+                    //if (TD.TcpConnectionResetErrorIsEnabled())
```

Comment:
are the commented code intentional or else you can remove

Reply:
Intentionally. At some point I need to add ETW events. The code is littered with TD.XXXXX commented code in anticipation as it will be easier to light up by leaving them in.